src/pdn/test/pdn_aux.py

```diff
@@ -0,0 +1,676 @@
+import re
+import utl
+import pdn
+import odb
+from collections import defaultdict
+
+
+#  In tcl land, this lives in OpenRoad.tcl. However, it seems to be only called
+#  in the pdn regression tests so it is defined here instead.
+def add_global_connection(design, *,
+                          net_name=None,
+                          inst_pattern=None,
+                          pin_pattern=None,
+                          power=False,
+                          ground=False,
+                          region=None):
+    if net_name is None:
+        utl.error(utl.PDN, 1501, "The net option for the " +
+                  "add_global_connection command is required.")
+
+    if inst_pattern is None:
+        inst_pattern = ".*"
+
+    if pin_pattern is None:
+        utl.error(utl.PDN, 1502, "The pin_pattern option for the " +
+                  "add_global_connection command is required.")
+
+    net = design.getBlock().findNet(net_name)
+    if net is None:
+        net = odb.dbNet_create(design.getBlock(), net_name)
+
+    if power and ground:
+        utl.error(utl.PDN, 1551, "Only power or ground can be specified")
+    elif power:
+        net.setSpecial()
+        net.setSigType("POWER")
+    elif ground:
+        net.setSpecial()
+        net.setSigType("GROUND")
+
+    # region = None
+    if region is not None:
+        region = design.getBlock().findRegion(region)
+        if region is None:
+            utl.error(utl.PDN, 1504, f"Region {region} not defined")
+
+    design.getBlock().addGlobalConnect(region, inst_pattern, pin_pattern, net, True)
```

Comment:
Lets not co-mingle things more. If this lived in OpenROAD land for tcl, I think it should live there as well for python.

Reply:
There are a couple of reasons why we might not want that. First is the fact that these ***_aux.py files are not really considered (yet) part of a defined Python API. I think @maliberty would like to have as much of the final API defined in C++ as possible, with the wrappers being very thin. I have mostly (but not always) been able to do that, but I have translated varying parts of the support tcl into Python (in the ***_aux.py files) in order to expediently port the regression tests to Python. The other reason to keep it here, for now, is avoid the "what namespace is that in?" problem.

Reply:
Most modules have and *_aux.py.  I view them all as temporary scaffolding until we refine the c++/python APIs.  My goal is that c++ & python have the same API and simple swig'ing is sufficient.  To make that a reality we'll have to refactor the APIs as well as move error checking into C++.  utl/include/utl/validation.h is a step in that direction.

In terms of process I want Don to get everything working and then I'll come through for API improvements.  He didn't feel familiar enough to refactor all the APIs involved.  @gadfort - you are welcome to handle this for pdn if you wish.

In general the python api will be more programmatic/oo than the command oriented tcl api.